### DISCUSSION TITLE
How to set a default value for a property.

### DISCUSSION THREAD

**Discussion by jonothedev** [spatie/laravel-data#416] [Q&A]
How can we set a default value for a property? For example, if i were to create a data object, but set a value for a property from two other properties..

```
public string $first_name
public string $last_name
```

Then set a default value for username

`$this->username = $this->first_name . '_' . $this->last_name;`

Is this possible?

**Comment by wivaku** [comment]
Support in Laravel Data is PHP's support for [simple regular class property defaults](https://stitcher.io/blog/constructor-promotion-in-php-8#simple-defaults), e.g.

```
class UserData extends Data
{
  public function __construct(
    public string $username = 'unknown',
  ) {}
```

For what you describe: duplicate of https://github.com/spatie/laravel-data/discussions/414

**Reply by linaspasv** [reply]
> Support in Laravel Data is PHP's support for [simple regular class property defaults](https://stitcher.io/blog/constructor-promotion-in-php-8#simple-defaults), e.g.
> 
> ```
> class UserData extends Data
> {
>   public function __construct(
>     public string $username = 'unknown',
>   ) {}
> ```
> 
> For what you describe: duplicate of #414

but what about if I want to load default values from lets say enum? Regular PHP does not allow assignments like these.

e.g. 

```php
#[MapName(SnakeCaseMapper::class)]
class Settings extends Data
{
    public function __construct(
        public array $acceptedDeliveryTypes = []     // I would like this to have defaults of OrderDeliveryType::values() fed from enum.
    ) {

    }
```

**Comment by Nir-An** [comment]
Hi @jonothedev  :) 

If the source of true is coming from the model , Should we need default property in the dto ? 

Why not added default values in the models inserted  ? 
https://laravel.com/docs/10.x/eloquent#default-attribute-values

**Comment by czuniga9** [comment]
Hi,
I've been looking at a similar issue and came out with adding a static function that massages the input into something the constructor is happy with, so for this example:

```php
class User extends Data
{
    public function __construct(
        public string $username,
    ) {}

    public static function createFromApiX(array $data): static
    {
        return static::from([
            'username' => $data['first_name'] . '_' . $data['last_name'],
        ]);
    }
}

$obj = User::createFromApiX([
    'first_name' => 'foo',
    'last_name'  => 'bar',
]);
assert('foo_bar' === $obj->username);
```

But I wonder if there's any drawbacks to this? One already is that I can't pass a json string directly like you can with `from` without more code. Another issue I found is that I can't use `fromApiX` as the method name, it seems to fail when the function starts with `from`.

**Reply by linaspasv** [reply]
> Hi, I've been looking at a similar issue and came out with adding a static function that massages the input into something the constructor is happy with, so for this example:
> 
> ```
> class User extends Data
> {
>     public function __construct(
>         public string $username,
>     ) {}
> 
>     public static function createFromApiX(array $data): static
>     {
>         return static::from([
>             'username' => $data['first_name'] . '_' . $data['last_name'],
>         ]);
>     }
> }
> 
> $obj = User::createFromApiX([
>     'first_name' => 'foo',
>     'last_name'  => 'bar',
> ]);
> assert('foo_bar' === $obj->username);
> ```
> 
> But I wonder if there's any drawbacks to this? One already is that I can't pass a json string directly like you can with `from` without more code. Another issue I found is that I can't use `fromApiX` as the method name, it seems to fail when the function starts with `from`.

Why not just use User::from directly? :)

**Reply by czuniga9** [reply]
Because I need to modify the array before passing it to User::from

**Comment by rubenvanassche** [comment]
I've added some extra docs on default values + a computed value support, should be landing later today.
